### ISSUE TITLE
[VS Code] JS/TS always selects the first suggestion

### ISSUE THREAD

**Issue by svipas** [MicrosoftDocs/intellicode#53]
**Product and Version** [VS/VSCode]: VS Code
**OS Version** [macOS/Windows]: macOS 10.14.3
**IntelliCode Extension Version**: 1.1.2
**Target Platform or Language** [e.g. Node.js]: JavaScript & TypeScript

**Steps to Reproduce / Scenario:**

1. Make sure you set `recentlyUsed` in `editor.suggestSelection` which is default in VS Code.
2. Disable this extension and try to use auto-complete and select second or any item from the list you want except the first one.
3. Again use auto-complete to render suggestions list on the same object and it should automatically point to your recently used suggestion.
4. Enable this extension and try to repeat 2 and 3 step. As you can see this extension disables it and acts like it's set to `first` in `editor.suggestSelection`.

I expect this feature to work properly with this extension enabled.

```
recentlyUsed: Select recent suggestions unless further typing selects one, e.g. `console.| -> console.log` because `log` has been completed recently.
```

Thank you!

**Comment by svipben** [comment]
Also, saw if I add `.` (dot) in the string I will get auto-complete which is not correct especially with such context:

<img width="677" alt="screenshot 2019-01-25 at 11 06 49" src="https://user-images.githubusercontent.com/10441177/51735911-54689e80-2091-11e9-8a71-10e17c25f66b.png">

**Comment by jkeech** [comment]
I believe the first issue is by design, since the IntelliCode extension moves the recommended items to the top of the completion list, and by default it selects the most-likely recommendation. In many cases (but not all), the most likely recommendation does match the recently used member.

I can also reproduce the issue with IntelliSense popping up inside the string literal, but only when coming back to it after completing the statement. When typing it out initially, IntelliSense does not pop up. Thanks for reporting this bug.

![js2](https://user-images.githubusercontent.com/746020/51763594-cb993500-2087-11e9-8f6c-c5747ce31a59.gif)

/cc @minestarks

**Comment by minestarks** [comment]
@jkeech who should "win" in [this case](
https://github.com/MicrosoftDocs/intellicode/issues/53#issuecomment-45801216)7? IntelliCode or VS Code? They're both doing their best to push most relevant suggestions to the top :smile:

(Even if the answer is ideally IntelliCode, I'm actually not sure if a plugin like IntelliCode can override the final sorting done by VS Code.)

@svipben Do you have an example for the behavior in your original description, i.e. where IntelliCode appears to override/disable the `recentlyUsed` option?

**Comment by jkeech** [comment]
@minestarks, I would expect IntelliCode suggestions to show up at the top if the extension is enabled and it offers suggestions. Do you know why the IntelliCode suggestions are normally sorted at the top, but in this case, VS Code pushed two suggestions higher?

@svipben, the extension recommends the most likely (popular) suggestion **based on the current context**. Since the context continually changes as you type, the suggestions can adapt. So for instance, if you are repeating the same member multiple times, and if that pattern is common among other codebases, then IntelliCode would also recommend that pattern in your case.

For example, if you start typing in a function without much context, IntelliCode might recommend one thing, but you choose a different path. It then adapts and recommends the next likely thing based on what you have done so far, even though it was different than the initial suggestion. It doesn't necessarily recommend the exact same members every time. Here's an example showing what I mean. Notice how the recommendations change as I go:

![intellicode-adapttocontext](https://user-images.githubusercontent.com/746020/51927773-3361bf80-23a9-11e9-8df4-76498645d414.gif)

**Comment by svipben** [comment]
Sorry for creating 3 issues in one. I separated `string literal` bug and `order` bug in two different issues.

**Comment by minestarks** [comment]
@svipben oh thanks, that's really helpful. So this one is just tracking the fact that IntelliCode appears to disable/ignore the `editor.suggestSelection` setting correct?

**Comment by svipben** [comment]
@minestarks Yeah, it looks like it doesn't work at all with IntelliCode. But I think you should carefully think of `recentlyUsed` because second time after you already selected something suggestions list will point to the recent one after typing `.` (dot) instead of showing suggestions list with IntelliCode suggestions at the top (like it's right now) and I kinda think it's correct behavior, **BUT** `recentlyUsedByPrefix` should work as expected which I mentioned above.

**So long story short:**
`recentlyUsed`: disable and keep like it's, select always first suggestion no matter what because we care about order and view of our IntelliCode suggestion list
`recentlyUsedByPrefix`: fix it as I mentioned above
`first`: works as supposed to 😄

**Comment by vivlimmsft** [comment]
Yes, the IntelliCode extension declares `configurationDefaults` for each language we support, which set `"editor.suggestSelection": "first"`.

You can override the setting by specifying a language-specific override setting for `editor.suggestSelection`. Could you please try the below, reload vscode, and try some suggestions?

```json
    "[javascript]": {
        "editor.suggestSelection": "recentlyUsedByPrefix"
    },
    "[typescript]": {
        "editor.suggestSelection": "recentlyUsedByPrefix"
    },
    "[java]": {
        "editor.suggestSelection": "recentlyUsedByPrefix"
    },
    "[python]": {
        "editor.suggestSelection": "recentlyUsedByPrefix"
    }
```

**Comment by svipben** [comment]
@vivlimmsft you're right, with this configuration it works. I understand why you kept `first` here, but maybe it's better to remove `configurationDefaults` and create a config like this:

```json
{
  "vsintellicode.suggestSelection": {
    "type": "string",
    "enum": [
      "first",
      "recentlyUsed",
      "recentlyUsedByPrefix"
    ],
    "default": "first",
    "description": "Controls how suggestions are pre-selected when showing the suggest list."
  }
}
```

Then user can decide which one is better for him.

**Comment by svipben** [comment]
Guys any news on this? I'm not pushing you though, but I'm interested if you're thinking about my suggestion to introduce that setting mentioned above. Thanks!

/cc @vivlimmsft @minestarks @markw-t @jkeech

**Comment by svipben** [comment]
@jkeech Thanks for response, so in all scenarios it would add these in my settings?

```json
    "[javascript]": {
        "editor.suggestSelection": "recentlyUsedByPrefix"
    },
    "[typescript]": {
        "editor.suggestSelection": "recentlyUsedByPrefix"
    },
    "[java]": {
        "editor.suggestSelection": "recentlyUsedByPrefix"
    },
    "[python]": {
        "editor.suggestSelection": "recentlyUsedByPrefix"
    }
```

**Comment by jkeech** [comment]
In the last scenario, it wouldn’t override your settings at all because you are already using a compatible setting at the global level.

**Comment by svipben** [comment]
I like the last scenario then. It checks if it's `recentlyUsedByPrefix` or `first` if yes it doesn't do anything, but if it's a `recentlyUsed` it will change it to a `first` by overriding it like mentioned above.

**Comment by svipben** [comment]
Also, you should consider this situation:

```json
"editor.tabCompletion": "on",
"editor.quickSuggestions": false,
```

In this situation, it should respect whatever `editor.suggestSelection` you have because you don't see any suggestions and you're completing them by pressing `tab`. So in this situation, even `recentlyUsed` is possible. Currently, I'm using it and I guess I will stick with it, because suggestion popup is too noisy and distracting, if I really don't know something I manually invoke suggestions. What do you think about this?

/cc @vivlimmsft @minestarks @markw-t @jkeech

**Comment by jkeech** [comment]
I think the general principle should be that we do not alter when the completion list shows up, but when it does show up, we ensure that the IntelliCode suggestions are the top/default options if you have the IntelliCode extension enabled.

In the scenario where you want different completion list sorting behavior that is conditioned on whether the list is visible or not -- I don't think this is possible without a VS code editor change. The sorting is something that the editor itself is doing, and as far as I know, it's not configurable based on when the completion list is being invoked, other than the per-language overrides. I think you have to pick one setting or the other, and it applies equally whether or not the list is visible. So in this case, it would still have to be `first` or `recentlyUsedByPrefix` so that if you manually invoke the completions, the IntelliCode suggestions are displayed properly.

Even if it was possible to conditionally change the sorting based on completion list visibility, I'm not sure that it's the best approach. The overall idea is that, on average, the top IntelliCode suggestion should be more accurate than the top `recentlyUsed` suggestion. So even if you are blindly selecting the first option without really looking, the IntelliCode suggestion should be correct more often, which would make users more productive.

**Comment by svipben** [comment]
@jkeech Sounds good, waiting for an update then! Thank you 🔥

**Comment by jkeech** [comment]
We found out that extensions cannot programmatically set language-specific setting overrides using the VS Code APIs, so we are going to change the implementation slightly. The end result should be the same for you though -- if you have `"editor.suggestSelection": "recentlyUsedByPrefix"` set, IntelliCode will not interfere with that.

Here's the new logic, which should be available soon in the next update of the extension (1.1.4):
1) The IntelliCode extension will provide a default value for `editor.suggestSelection`, but not the language-specific overrides that it currently provides in 1.1.3. We will continue to use `first` as the default value, but if we hear enough feedback from users that they would generally prefer `recentlyUsedByPrefix`, we can always update to that as the default.
```json
"configurationDefaults": {
    "editor.suggestSelection": "first"
},
```
2) Since this is only a default, if you have explicitly set your own value for `editor.suggestSelection`, it will win. The order of precedence is user-defined settings (at the workspace folder, workspace, or global level), extension-provided defaults, VS Code-provided defaults.
3) With this logic, a user could explictly set a value that is incompatible with IntelliCode. During extension activation, we will check the current setting value and print a warning to the output window if the setting that is being used will interfere with the expected behavior of IntelliCode, but we will not automatically change the setting to a different value. This gives users full control. We try to provide good defaults that work well with IntelliCode, but users can always override it if they want.

**Comment by vivlimmsft** [comment]
We changed the implementation from what @jkeech mentioned above, because we found that [configurationDefaults](https://code.visualstudio.com/api/references/contribution-points#contributes.configurationDefaults) declared in `package.json` can **only** be applied to specific languages.

The new implementation, which is in 1.1.4, removes all defaults from our `package.json` and instead will apply `'editor.suggestSelection': 'first'` as needed, under the following conditions:

* If `editor.SuggestSelection` has not been manually configured to be `recentlyUsed`, and is getting that value by default, we will write `'editor.suggestSelection': 'first'` without asking. This replicates previous behavior, with the side effect of meaning that you cannot use `recentlyUsed` with any language unless you use a [language-specific setting](https://code.visualstudio.com/docs/getstarted/settings#_language-specific-editor-settings) to do so.
* If `editor.SuggestSelection` was manually configured by you to be `recentlyUsed`, we will show a notification asking if you'd like it to be changed to `first` automatically. You can also decline, in which case we will not prompt again.

We introduced a new setting, `vsintellicode.modify.editor.suggestSelection`, which keeps track of:
* whether you have declined to have `editor.suggestSelection` updated by IntelliCode.
* whether we have updated `editor.suggestSelection` on your behalf, and if so, which of the above conditions was true at the time. We're keeping track of those conditions in case we need to change this behavior in the future.

To opt out of this behavior, you can add `"vsintellicode.modify.editor.suggestSelection": "disabled",` to your `settings.json`. 

If you still want to use `recentlyUsed` with specific languages, you can set [language-specific settings](https://code.visualstudio.com/docs/getstarted/settings#_language-specific-editor-settings). For example:
```json
    "[csharp]": {
        "editor.suggestSelection": "recentlyUsed"
    },
```

**Comment by svipas** [comment]
Thanks, guys for the update. I'm really happy with these changes and I understand the reason. Good job!

**Comment by jkeech** [comment]
Version 1.1.4 has been published to the marketplace. Please give it a shot and let us know what you think. Thanks!
